mr/src/main/java/org/apache/iceberg/mr/mapred/IcebergSerDe.java

```diff
@@ -0,0 +1,129 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.mr.mapred;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.sql.Date;
+import java.sql.Timestamp;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+import javax.annotation.Nullable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.serde2.AbstractSerDe;
+import org.apache.hadoop.hive.serde2.SerDeException;
+import org.apache.hadoop.hive.serde2.SerDeStats;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
+import org.apache.hadoop.io.Writable;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.SnapshotsTable;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+public class IcebergSerDe extends AbstractSerDe {
+
+  private Schema schema;
+  private ObjectInspector inspector;
+  private List<Object> row;
+
+  @Override
+  public void initialize(@Nullable Configuration configuration, Properties serDeProperties) throws SerDeException {
+    Table table = null;
+    try {
+      table = TableResolver.resolveTableFromConfiguration(configuration, serDeProperties);
+    } catch (IOException e) {
+      throw new UncheckedIOException("Unable to resolve table from configuration: ", e);
+    }
+    this.schema = table.schema();
+    if (table instanceof SnapshotsTable) {
+      try {
+        this.inspector = new IcebergObjectInspectorGenerator().createObjectInspector(schema);
+      } catch (Exception e) {
+        throw new SerDeException(e);
+      }
+    } else {
+      List<Types.NestedField> columns = new ArrayList<>(schema.columns());
+      columns.add(Types.NestedField.optional(Integer.MAX_VALUE,
+          SystemTableUtil.snapshotIdVirtualColumnName(serDeProperties), Types.LongType.get()));
+      Schema withVirtualColumn = new Schema(columns);
+      try {
+        this.inspector = new IcebergObjectInspectorGenerator().createObjectInspector(withVirtualColumn);
+      } catch (Exception e) {
+        throw new SerDeException(e);
+      }
+    }
+  }
+
+  @Override
+  public Class<? extends Writable> getSerializedClass() {
+    return null;
+  }
+
+  @Override
+  public Writable serialize(Object o, ObjectInspector objectInspector) {
+    return null;
+  }
+
+  @Override
+  public SerDeStats getSerDeStats() {
+    return null;
+  }
+
+  @Override
+  public Object deserialize(Writable writable) {
```

Comment:
Instead of using ObjectInspectors, do you think it's worth building a ReaderFunc which returns Hive compatible objects instead? We can save one conversion that way. e.g Avro objects -> Hive objects, instead of Avro objects -> Iceberg Generic objects -> Hive objects. The downside is we will have to build a Hive-compatible ReaderFunc for each file format but I don't expect it to be too hard.

Reply:
+1. We should build this similar to reader Funcs. Please see examples of this in Spark module org.apache.iceberg.spark.data.SparkValueReaders [for Avro] and org.apache.iceberg.spark.data.SparkOrcValueReaders

Reply:
"There's actually no need to return anything other the Record object."

@guilload are you sure about this? Ultimately this all ends up being used by the old Map Reduce V1 API where serialization required the usage of Writables. Have you managed to get an InputFormat working end to end in an actual Hive 2.3.x client with a type that isn't a Writable? I see that the InputFormats etc. for Delta and Hudi both seem to use `ArrayWritable`and have Writables scattered over the code which I'm assuming is for this reason. I'd be happy to remove the Writable I'm just not sure it would actually work when run on a cluster.

Reply:
> Ultimately this all ends up being used by the old Map Reduce V1 API where serialization required the usage of Writables

I think the suggestion was to return `Record` from `deserialize`, not to use a non-writable. After `deserialize`, I don't think there is a requirement for anything to be a `Writable`.

Reply:
> Instead of using ObjectInspectors, do you think it's worth building a ReaderFunc which returns Hive compatible objects instead?

This seems like a larger effort that is probably something for a follow-up instead. It seems much easier to me to add a small translation layer just for the types that need it and use the current Iceberg generics, rather than building a new object model for data formats.

Reply:
> > Ultimately this all ends up being used by the old Map Reduce V1 API where serialization required the usage of Writables
> 
> I think the suggestion was to return `Record` from `deserialize`, not to use a non-writable. After `deserialize`, I don't think there is a requirement for anything to be a `Writable`.

We can certainly try, we should probably do it in Hiveberg first as we can then run a full HiveRunner test over it to make sure there aren't any surprises in the internals of Hive or MR.

Reply:
> > Ultimately this all ends up being used by the old Map Reduce V1 API where serialization required the usage of Writables
> 
> I think the suggestion was to return `Record` from `deserialize`, not to use a non-writable. After `deserialize`, I don't think there is a requirement for anything to be a `Writable`.

Exactly. Until getting to `deserialize` the object has to be a `Writable` but after that point, it can be anything really, as long as the right object inspector is provided. I have already successfully tested this approach on local branch where I built a simple Hive input format on top of #1104 and implemented my own SerDe with [this object inspector](https://github.com/guilload/iceberg/commit/3bffa7c87bf6352d5f7832edac383310aaf1df08#diff-0819121b2ca0adb7dfc72ceb03dd97b5).

Reply:
OK, sounds good, do you want to make this change as part of what you mention above with the Object Inspectors or later?

Reply:
Above changes have now been merged into this PR.